plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMergeSink.java

```diff
@@ -13,137 +13,74 @@
-        ImmutableList.Builder<Type> columnTypesBuilder = ImmutableList.builder();
-        columnNamesBuilder.addAll(phoenixOutputTableHandle.getColumnNames());
-        columnTypesBuilder.addAll(phoenixOutputTableHandle.getColumnTypes());
-        if (hasRowKey) {
-            columnNamesBuilder.add(ROWKEY);
-            columnTypesBuilder.add(ROWKEY_COLUMN_HANDLE.getColumnType());
-        }
-
-        PhoenixOutputTableHandle updateOutputTableHandle = new PhoenixOutputTableHandle(
-                remoteTableName,
-                columnNamesBuilder.build(),
-                columnTypesBuilder.build(),
-                Optional.empty(),
-                Optional.empty());
-        return new JdbcPageSink(session, updateOutputTableHandle, phoenixClient, pageSinkId, remoteQueryModifier);
-    }
-
-    private ConnectorPageSink createDeleteSink(
-            ConnectorSession session,
-            List<Type> mergeRowIdFieldTypes,
-            PhoenixClient phoenixClient,
-            ConnectorPageSinkId pageSinkId,
-            RemoteQueryModifier remoteQueryModifier)
-    {
-        checkArgument(mergeRowIdFieldNames.size() == mergeRowIdFieldTypes.size(), "Wrong merge row column, columns and types size not match");
-        JdbcOutputTableHandle deleteOutputTableHandle = new PhoenixOutputTableHandle(
-                remoteTableName,
-                mergeRowIdFieldNames,
-                mergeRowIdFieldTypes,
-                Optional.empty(),
-                Optional.empty());
-
-        return new DeleteSink(session, deleteOutputTableHandle, phoenixClient, pageSinkId, remoteQueryModifier);
-    }
-
-    private class DeleteSink
-            extends JdbcPageSink
+    public PhoenixMergeSink(PhoenixClient phoenixClient, RemoteQueryModifier remoteQueryModifier, ConnectorSession session, ConnectorMergeTableHandle connectorMergeTableHandle, ConnectorPageSinkId pageSinkId, QueryBuilder queryBuilder)
     {
-        public DeleteSink(ConnectorSession session, JdbcOutputTableHandle handle, JdbcClient jdbcClient, ConnectorPageSinkId pageSinkId, RemoteQueryModifier remoteQueryModifier)
-        {
-            super(session, handle, jdbcClient, pageSinkId, remoteQueryModifier);
+        PhoenixMergeTableHandle mergeTableHandle = (PhoenixMergeTableHandle) connectorMergeTableHandle;
+        PhoenixOutputTableHandle insertHandle = mergeTableHandle.insertHandle();
+
+        this.columnCount = insertHandle.getColumnTypes().size();
+        this.hasRowKey = insertHandle.rowkeyColumn().isPresent();
+
+        this.insertSink = new JdbcPageSink(session, insertHandle, phoenixClient, pageSinkId, remoteQueryModifier, Optional.empty());
+        this.updateSink = new JdbcPageSink(session, mergeTableHandle.updateHandle(), phoenixClient, pageSinkId, remoteQueryModifier, Optional.empty());
+
+        try (Connection connection = phoenixClient.getConnection(session)) {
+            String deleteSql = queryBuilder.prepareDeleteQuery(
```

Comment:
Maybe usage of `queryBuilder` could be extracted to separate commit, to make it easier to follow the changes.